### ISSUE TITLE
Taskbar order isn't preserved when turning off Nvidia surround

### ISSUE THREAD

**Issue by RBZL** [terrymacdonald/DisplayMagician#78]
I have two profiles set up on a system with a Nvidia graphics card:

- A normal triple monitor profile without Nvidia surround ("Triple Monitor")
- A Nvidia surround profile ("Surround")

![Untitled](https://user-images.githubusercontent.com/6739054/149615245-ecf5be5b-f896-4250-8146-232abe937556.png)

When switching from the Surround profile to the Triple Monitor profile, the primary taskbar with the system tray ends up on the right-most monitor. It is normally on the center monitor, and this is how the Triple Monitor profile was saved. I have to unlock the taskbars and drag the right taskbar to the center monitor to resolve the issue after switching from Surround to Triple Monitor.

Version 2.1.5.5, Windows 10. I can attach support log info if required. Thanks!

**Comment by terrymacdonald** [comment]
@RBZL Sorry, but that's something I can't control. Windows decides where it wants to put the taskbar, and there's not much I can do about. Sometimes it's been deciding lately to put it up the top of the left-most triple screen on my simrig, and it won't allow me to change the position until I reboot. Windows 10 seems to be much better at being reliable with the taskbar placement, but Windows 11 is not as nice. It seems to move about a LOT.

I am hoping to one day be able to discover how to force windows to put the taskbar where I want it to go, but at present I am unable to get this to work, as it's completely not supported by the Windows API.

Unfortunately I'll have to close this issue and mark it with 'cant fix', as it's out of my hands :(.

Thanks
Terry

**Comment by terrymacdonald** [comment]
Can't fix this as its a windows setting that is not accessible reliably via Windows API at present.

**Comment by RBZL** [comment]
Interesting - I recently upgraded to 2.1.5.5 from the 1.x series, and the issue wasn't present before. I'll look into it and advise if I can figure out if that's still the case and if I find anything else for adjusting the taskbars programmatically.

**Comment by terrymacdonald** [comment]
Hmmmm. That's an interesting data point. The 1.x series uses the
NvAPIWrapper library to manipulate the NVIDIA NvAPI. The 2.x series uses
the NVIDIALibrary written by me to do the manipulation. I wrote my own
library to be able to handle things like HDR, bevel spacing and refresh
rates.

I might take a look at how NvAPIWrapper handles seeing the display layout
to compare and see if there are any differences.

Terry

**Comment by terrymacdonald** [comment]
I've had a look at some of the taskbar setting logic in HeliosDisplayManagement that Soroush was working on, to see if I can use some of his awesome ideas. As is typical, Soroush does have some nice code, but it doesn't appear to be plugged in to the main codebase there. So no idea if it works. Additionally, win 11 actually restricts the user from making changes to the taskbar layout, so it may not actually allow changes to take place. I need to do some more testing with Win 10 as well to be sure it actually works.

That said, I'm trying to do some experiments to see what settings will work in Win11, and to see if I can at least save the taskbar position with each display profiles and then set it later. At this point it seems to work, but I need to spend a lot more time on it before I'll be happy it works reliably enough to include it in the main codebase. At the moment it is in it's own feature branch here: https://github.com/terrymacdonald/DisplayMagician/tree/feature-taskbar-setting

I'll look at getting people to test the code for me once I finish doing my own testing, and probably after I've fixed some of the NVIDIA Surround issues that a small percentage of people are getting.

**Comment by RBZL** [comment]
Sorry, I had notifications turned off for some reason. Cool, if you need any other information from me or help with testing I'm happy to help out.

**Comment by terrymacdonald** [comment]
Well, I've had some successes! I can now store and reset the Win10 taskbar, and I am closer to doing the same thing on Windows 11 (It's a pain on Win 11!). It won't help with your particular problem though I don't think, as I can't figure out how to tell windows which screen is the main one to show the taskbar on. I'll try and do a bit more digging, but there doesn't look to be much prior work on this part :(.

I've read that Microsoft is looking at allowing people to move the taskbar location in an upcoming version of Win11, so my code is likely to start working again once Microsoft officially starts allowing users to move the taskbar (rather than an unofficial reg hack).

Thanks
Terry

**Comment by terrymacdonald** [comment]
OK - some further good news. I have figured out a whole list of what the different Windows taskbar settings do, and I have managed to modify some code from Soroush to specifically get these settings (dropping here for the internet to find later on if ppl need it):

```
             // Win10/11 registry keys (not all will be populated, only those that the user modified from default at least once)
            "MMTaskbarEnabled", //  Multiple Taskbars:  0 for show taskbar on main screen only, 1 for show taskbar on all screens            
            "MMTaskbarMode", // Show taskbar buttons on: 0 = all taskbars, 1 = main taskbar and where windows is open, 2 = taskbar where window is open
            "MMTaskbarGlomLevel", // Buttons on other taskbars: 0 = always combine, combine when the taskbar is full, 2 = never combine
            "NoTaskGrouping", // Disable all Task Grouping (overrides "TaskbarGroupSize"): 0 = enable task grouping, 1 = disable task grouping
            "SearchboxTaskbarMode", // Show Search Button in Taskbar: 0 = remove search button, 1 = show search button
            "ShowTaskViewButton", // Show Taskview Button in Taskbar: 0 = remove taskview button, 1 = show taskview button
            "TaskbarAl", // Start Button Alignment: 0 for left, 1 for center, 
            "TaskbarDa", // Show Widgets button in Taskbar: 0 = remove widgets button, 1 = Show widgets button
            "TaskbarGlomLevel", // Buttons on main screen: 0 = always combine, combine when the taskbar is full, 2 = never combine
            "TaskbarGroupSize", // TaskBar left/right grouping by age: 0 = oldest first (default), 1 = roup by size largest first, 2 = group all with 2 or more, 3 = group all with 3 or more (see NoTaskGrouping to prevent Grouping )
            "TaskbarMn", // Show Chat Button in Taskbar: 0 = remove chat button, 1 = show chat button
            "TaskbarSi", // Taskbar Size: 0 = small, 1 = medium, 2 = Large
            "TaskbarSizeMove", // Lock the Taskbar (prevent resizing): 0 = taskbar size is locked, 1 = taskbar size is unlocked
            "TaskbarSmallIcons", // Small Taskbar Icons: 0 = normal sized icons, 1 = small icons
            "TaskbarSd", // Show Desktop Button in Taskbar: 0 for hid the show desktop button, 1 for show the Show desktop button
```

This will hopefully mean that future versions of DisplayMagician will be able to make sure the right taskbars are in the right places 🤞.

This will, however, require that DisplayMagician patches all the existing configs that people have. The best I'll be able to do is to patch the older version of the config files with the current taskbar settings, and then if those aren't right people will have to recreate the display layouts again exactly the way they want them.

I'll hopefully have something for you to test in a little while. Are you still on Windows 10? Win 11 seems to require an explorer restart to make the taskbar move, which is proving more difficult to code than it should be :).

Thanks
Terry

**Comment by terrymacdonald** [comment]
OK - I've made something that kinda works in Win 11 (registry changes take place but explorer ignores the main taskbar move :( ) but is highly likely to work in Windows 10.



Can you please install DisplayMagician v2.2.0.98 test version here and see if you can make new taskbar settings? I would like you to please try making new Display Profiles that:
- Try the taskbars in different edges of the screen
- Try it will only one taskbar
- Other features like that you can think of.

[DisplayMagicianSetup-v2.2.0.98.zip](https://github.com/terrymacdonald/DisplayMagician/files/7924171/DisplayMagicianSetup-v2.2.0.98.zip)

Please note, I'm not sure that the DisplayMagician can detect the hidden taskbar yet, as I'm not sure where that setting is stored in registry, but most other settings should now be tracked.

Please give it a go, and tell me what you think.

Thanks
Terry

**Comment by terrymacdonald** [comment]
That surround screen issue where it showed 3 screens but has one was odd.
The left over background menu is because I forgot to update it to the new
config (which seems to be broken).

It sounds like the last bit meant that DisplayMagician now has an invalid
config and won't start. I suspect it is because you used the desktop
connect menu which unfortunately was pointing to the old invalid config.

Can you please ZIP up the contents of
C:\Users\<username>\AppData\Local\DisplayMagician and attach it to this
issue? Id really like to see exactly where it is crashing by looking at the
logs and checking out the configuration files. I suspect it is when it is
trying to use the taskbar settings and is crashing when they are missing.

I should be able to give you a working config back at least, but the
taskbar part of it sounds like it still needs some work.

Thanks
Terry

**Comment by RBZL** [comment]
[DisplayMagician.zip](https://github.com/terrymacdonald/DisplayMagician/files/7945906/DisplayMagician.zip)

I'll nuke the contents of AppData\Local\DisplayMagician and play with it some more. Setting up the profiles again isn't too difficult.

**Comment by terrymacdonald** [comment]
@RBZL Great news! I dived into the Windows internal messages and started monitoring the way that the Windows 10 and Windows 11 Settings app changes the taskbar placement instantly, and I managed to replicate it! \o/ I've not seen this documented anywhere on the internet, so I think I may have discovered something new :).

For reference for people in the internet, this is the code I ended up with: 

```
[DllImport("user32.dll", SetLastError = true, CharSet = CharSet.Auto)]
public static extern IntPtr FindWindow(string lpClassName, String lpWindowName);

[DllImport("user32.dll", SetLastError = true, CharSet = CharSet.Auto)]
public static extern IntPtr FindWindowEx(IntPtr hwndParent, IntPtr hwndChildAfter, string lpszClass, string lpszWindow);

[DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Auto)]
public static extern int SendMessage(IntPtr hWnd, uint wMsg, IntPtr wParam, IntPtr lParam);

public const int NULL = 0;
public const int WM_SETTINGCHANGE = 0x001a;
public const int SPI_SETWORKAREA  = 0x002F;
public const int WM_USER_REFRESHTASKBAR = 0x05CA;
public const int wParam_SHELLTRAY = 0x00000006;

public static bool RepositionMainTaskBar(TaskBarEdge edge)
        {
            // Tell Windows to refresh the Main Screen Windows Taskbar
            // Find the "Shell_TrayWnd" window 
            IntPtr mainToolBarHWnd = Utils.FindWindow("Shell_TrayWnd", null);
            // Send the "Shell_TrayWnd" window a WM_USER_REFRESHTASKBAR with a wParameter of 0006 and a lParamater of the position (e.g. 0000 for left, 0001 for top, 0002 for right and 0003 for bottom)
            IntPtr taskBarPositionBuffer = new IntPtr((Int32)edge);
            Utils.SendMessage(mainToolBarHWnd, Utils.WM_USER_REFRESHTASKBAR, (IntPtr)Utils.wParam_SHELLTRAY, taskBarPositionBuffer);
            return true;
        }

        public static bool RepositionSecondaryTaskBars()
        {
            // Tell Windows to refresh the Other Windows Taskbars if needed
            IntPtr lastTaskBarWindowHwnd = (IntPtr)Utils.NULL;
            for (int i = 0; i < 100; i++)
            {
                // Find the next "Shell_SecondaryTrayWnd" window 
                IntPtr nextTaskBarWindowHwnd = Utils.FindWindowEx((IntPtr)Utils.NULL, lastTaskBarWindowHwnd, "Shell_SecondaryTrayWnd", null);
                if (nextTaskBarWindowHwnd == (IntPtr)Utils.NULL)
                {
                    // No more windows taskbars to notify
                    break;
                }
                // Send the "Shell_TrayWnd" window a WM_SETTINGCHANGE with a wParameter of SPI_SETWORKAREA
                Utils.SendMessage(lastTaskBarWindowHwnd, Utils.WM_SETTINGCHANGE, (IntPtr)Utils.SPI_SETWORKAREA, (IntPtr)Utils.NULL);
                lastTaskBarWindowHwnd = nextTaskBarWindowHwnd;
            }
            return true;
        }
```

The upshot of all this is that I'm now making some changes to the way that the Taskbar settings are stored, as well as adding the ability to change the position of the Taskbars from within DisplayMagician (even on Windows 11!). Please note that Win 11 taskbars won't work on the left or right at the moment, only top and bottom).

I hope to have something sorted soon!

Thanks
Terry

**Comment by RBZL** [comment]
Sounds good, thanks for the work. I'm happy to test it out once you've got something built.

**Comment by terrymacdonald** [comment]
@RBZL I've just finished getting taskbar settings to record and I'm pretty happy with how they turned out :). I've also configured DisplayMagician to ignore getting HDR/Colour settings from the following types of connections: VGA, COMPONENT, SVIDEO, DVI, COMPOSITE. 

That should stop you getting errors (hopefully), so it would be cool if you could both install DisplayMagician v2.2.0.179 (yes this is build number 179!) that I've attached here.

IMPORTANT: YOU WILL NEED TO DELETE YOUR DisplayProfiles_2.2.json file AND CREATE NEW PROFILES FROM SCRATCH. If you want to back up your current settings before you do so, you can copy your C:\Users\<username>\AppData\Local\DisplayMagician folder (or make a zipped up copy) so that you don't lose any settings. Please note that installing this test version may mean you will still need to do this again once the final v2.2.0 is released.

This version includes changed DisplayProfile format again in order to track all the new features we've been working on recently. It will require new display profiles.

[DisplayMagicianSetup-v2.2.0.179.zip](https://github.com/terrymacdonald/DisplayMagician/files/8001419/DisplayMagicianSetup-v2.2.0.179.zip)

Please test this out and see if there are any broken bits. If there are, please upload a Support ZIP File with a description of what's wrong. 

I think we're getting really close to a new release!

Thanks
Terry

**Comment by RBZL** [comment]
I installed 179 and did a little bit of testing tonight. I had various issues so to start fresh I did the following:

- Exited DisplayMagician entirely
- Deleted the \Local\DisplayMagician folder to reset everything
- Started it back up to play with it

I did this in my normal triple monitor habitat with no NVIDIA surround enabled to start with a clean slate.

I saved my profile as "Triple Monitor 4" so it was a unique new name. Then I set up and enabled NVIDIA surround, and saved that as "Surround 4". I left taskbars and the like where they're at on the bottom of the screen, with center being primary.

I switched back to "Triple Monitor 4" using DisplayMagician, and my primary taskbar was on the left screen even though it had been on the center when the profile was saved and the center monitor was set as primary. I deleted the "Triple Monitor 4" profile, fixed the taskbars, and saved it as "Triple Monitor 5". I switched back to the "Surround 4" profile and surround activated successfully, but my taskbar spans all monitors and DisplayMagician crashed. I was able to re-open DisplayMagician, so I created a support file:

[DisplayMagician-Support-20220204-2157.zip](https://github.com/terrymacdonald/DisplayMagician/files/8007413/DisplayMagician-Support-20220204-2157.zip)

I re-activated "Triple Monitor 5" and everything is as expected. I activated "Surround 4" again and it worked, and the taskbar stayed on the center screen only and DisplayMagician didn't crash.

I recalled that when the primary taskbar ended up on the left monitor, the DisplayMagician application was on the left monitor too when I applied the "Triple Monitor 4" profile. While in "Surround 4", I moved DisplayMagician's application windows to the left monitor and activated "Triple Monitor 5" - recall that this just worked successfully when DisplayMagician was on the center monitor above in "Surround 4" before I activated "Triple Monitor 5". With DisplayMagician on the left monitor, that monitor also received the primary taskbar. I reactivated "Surround 4" with DisplayMagician on the left monitor and it crashed again. I opened DisplayMagician again and created a support file:

[DisplayMagician-Support-20220204-2206.zip](https://github.com/terrymacdonald/DisplayMagician/files/8007425/DisplayMagician-Support-20220204-2206.zip)

So it seems that if I activate NVIDIA surround while DisplayMagician isn't on the center/primary monitor, it will crash, and if I activate my normal triple monitor setup which disables NVIDIA surround while DisplayMagician isn't on the center/primary monitor, the monitor with DisplayMagician on it will get the primary taskbar. I tested this by moving DisplayMagician to the right monitor to see if it would get the primary taskbar and switched to "Triple Monitor 5", and it did not - the center monitor got the primary taskbar (as saved in that profile) and DisplayMagician application windows ended back up on the primary monitor after the switch.

Note that two of my monitors, the center and right, are connected via DisplayPort, and the left is connected via DVI. They are all connected to one single GTX 1070. So it seems the DisplayPort monitors may not cause issues but the DVI may.

Also, at some point in time, my desktop icon positions were reset for the "Triple Monitor" profile. Previously, they stayed where they were throughout profile switches. I believe before I reset the \Local\DisplayMagician folder, they were consolidated to the center monitor in one big blob after one of the profile switches where DisplayMagician crashed.

Here is the final support file after all of the above:

[DisplayMagician-Support-20220204-2211.zip](https://github.com/terrymacdonald/DisplayMagician/files/8007430/DisplayMagician-Support-20220204-2211.zip)

Prior to resetting my config directory when I started having problems, I did try moving all taskbars to the left on all monitors for Triple Monitor, and with a switch to Surround and Back they stayed in place.

Also, I noticed that after switching back to Triple Monitor from Surround, if I enter the NVIDIA control panel to view the surround settings, the "Enable Surround" box is always checked even if surround is currently disabled.

Finally, I noticed that the monitor order is odd between surround enabled and disabled. If surround is disabled, the monitors as they are laid out physically are 3 2 1 (both in NVIDIA settings and Windows display settings). For surround to be enabled and the monitors to be in the correct order, they must be 1 2 3 in the NVIDIA surround settings.

I know the above might result in more questions than answers, but hopefully something useful is in there. Let me know what you need next.

**Comment by terrymacdonald** [comment]
Woah. Thanks for the detailed notes. It looks like there is a bit to unpick
in there. I'll take a few days to check through what you've sent and I'll
get back to you with any questions.

Thanks so much for all your testing!

Terry

**Comment by terrymacdonald** [comment]
@RBZL OK! I think that the DVI connection is the ultimate cause of your issue, though there could some other additional problems in your logs that I missed.

I fixed the DVI issue yesterday, and another user said that it worked fine on their test system. I have also spent the day working through some bugs with the drawing of where the taskbars are, and I think I have that fixed too!

![image](https://user-images.githubusercontent.com/5721772/152664225-ebb30318-e255-40d4-8460-1c968acaa59b.png)

Can I please trouble you to do some testing for me again? I *think* I have fixed the issues you mentioned , but I really need confirmation from you that it is now working.

IMPORTANT: Just to be sure, can you delete all your C:\Users<username>\AppData\Local\DisplayMagician\Profiles\DisplayProfiles_*.json files just so we know that all the configs are generated in the right config format? This will let me make sure it is working in a known state and avoid some of the past bugs.

Then please download DisplayMagician v2.2.0.223 and install it. Then set up new Display Profiles, and then use those profiles in the Game Shortcuts. 
[DisplayMagicianSetup-v2.2.0.223.zip](https://github.com/terrymacdonald/DisplayMagician/files/8009215/DisplayMagicianSetup-v2.2.0.223.zip)
 
Please test the following:
- Test changing between surround and non-surround profiles
- Please test changing taskbar position (can do that using the new 'Profile Tools' button under Profiles.
- Please check the UI shows the correct display layout and that the taskbar layout in the UI matches the real thing.

And then please report back with what you've found.

Thanks
Terry

**Comment by RBZL** [comment]
I'll try to check this out tomorrow, thanks.

**Comment by RBZL** [comment]
I played around with it a bit and encountered a number of issues. To write them up better, I deleted the \Local\DisplayMagician folder, reset my taskbars to the bottom with surround disabled, and opened DisplayMagician to start again.

I went to save the current profile, and the monitor previews show the center taskbar at the top of the screen and the side taskbars on the bottom. All taskbars are actually located at the bottom of the screens. I used Profile Tools to move the taskbars to the bottom, which fixed the preview. The taskbars didn't actually move since they were already at the bottom. I saved this as TripleBoi.

I set up Surround and saved that profile as SurroundBoi. I activated TripleBoi and my primary taskbar ended up on the left monitor with the DVI connection. I manually dragged it back to center, deleted TripleBoi, saved TripleBoi again, and switched to SurroundBoi. The app crashed, so I opened it again. I switched to TripleBoi and the primary taskbar was on the left screen again. I dragged it back to center, didn't save anything, and switched to SurroundBoi (app didn't crash this time) and back again. The primary taskbar was on the center screen for an instant when the screens were in transition, but ultimately it keeps ending up on the left monitor now. I created a support file:

[DisplayMagician-Support-20220209-2041.zip](https://github.com/terrymacdonald/DisplayMagician/files/8037870/DisplayMagician-Support-20220209-2041.zip)

I used Profile Tools to move the taskbars to each side of the screen, and that seemed to work. The preview images updated accordingly. The primary taskbar stayed on my center monitor, but on whatever side I told the taskbars to move to.

I put the taskbars on top of the screens and saved it as TripleBoiTop. I activated SurroundBoi and then TripleBoiTop, and the taskbars ended up on top but the primary was on the left monitor again. I activated TripleBoi and the taskbars ended up on the left of the screens somehow - they were never saved that way. The Display Profiles window showed the taskbar previews on the left, but appeared as if a new profile was ready to be saved instead of showing TripleBoi active. The TripleBoi previews still show the taskbars on the bottom. I tried activating TripleBoi again, and after a flicker there was no change in position but the TripleBoi profile now appears to be active even though the taskbars are on the left instead of the bottom. I went back and forth between TripleBoi and TripleBoiTop a few times, and TripleBoiTop always puts them on top. About 50% of the time activating TripleBoi results in the correct preview of the taskbars on the left (even though they shouldn't be there) and the option to save a new profile, and 50% it results in TripleBoi showing as active but the taskbars being on the left. I created a support file:

[DisplayMagician-Support-20220209-2054.zip](https://github.com/terrymacdonald/DisplayMagician/files/8037928/DisplayMagician-Support-20220209-2054.zip)

Other notes:

Profile management is not completely intuitive. It would be nice to have a way to overwrite profiles with the current settings (unless I'm missing something and that exists). If I move the taskbar position and just want to save that, the selected profile only gives me the option to rename it, not to save the current settings under the same name. If I click Rename with the same name, I get an error. I also don't like that if you click an existing profile, then click away from it to deselect it, you can't save another profile under a new name. Sometimes I'll be looking at a profile, make a change, and want to save that change, but I can't without closing the Display Profiles window and opening it again.

If I deselect all profiles in Display Profiles by clicking away from them, I still have the option to apply or delete what appears to be the last selected profile. If I click Delete, I cause an unhandled exception. If I click Apply, it seems to apply the last selected profile. I created a support file after causing the exception a few times and putting my taskbars back to the bottom:

[DisplayMagician-Support-20220209-2058.zip](https://github.com/terrymacdonald/DisplayMagician/files/8037946/DisplayMagician-Support-20220209-2058.zip)

The option to open DisplayMagician to the notification area should probably be enabled by default. I find that otherwise when I close the application window, it closes completely and is annoying to have to wait for it to re-launch each time.

**Comment by terrymacdonald** [comment]
Thanks @RBZL,

Yeah, I found some issues with 2.2.0.223 not picking up the taskbar
position properly. I think that's the root of your issue with everything
moving to the left. Microsoft made Left = 00, and of course when it
WinLibrary fails to pick up the taskbar position properly it defaults to
0...which is Left.

I've also noted the problem with the profile management. This is actually a
bug in the equality matching logic for the taskbar. The idea was always
that DisplayMagician records your settings as they are...it's a display
layout recorder/player. The idea would be that you would use something to
move your taskbars around and change your display layout, and then when
you're ready you would take a snapshot of the layout and DisplayMagician
would store all the settings for later. It would make things exactly the
same when you applied the saved profile later.

But, I thought that changing the taskbar position was a bit too difficult
for people to do, especially in Windows 11. So I made the 'Profile Tools'
button, which opens the Profile Tools window, and that contains a taskbar
position changer tool. This tool does not affect the saved display profiles
at all, but is instead a tool that will make the change right now to the
display layout. It will then allow you to save the changed display layout
as a new display profile. In other words, you setup the layout you want,
then you go into DM Profile Tools and change the taskbar position, and then
you're ready to save the Display Profile.

That was the idea anyway, but now I think it might be better to actually
remove that profile tools button as it will just confuse people. I am now
thinking it would be easier to explain to people if DisplayMagician only
records and applies display Profiles, and doesn't allow you to do any
'real-time' changes at all itself (like the changing taskbar position right
now).

What do you think? Does having Profile Tools send a mixed message?

I know many people don't like having DisplayMagician run on boot, so I will
leave it off by default. I personally always make sure it's on as it's so
much faster now that it's a single instance (there was a LOT of programming
to make that happen!).

Thanks

*Terry *

**Comment by terrymacdonald** [comment]
Oh I forgot to mention, I should have another version of DM for you to test
in a few days. I'm working through the taskbar bugs as I find them and
trying to make something reliable that just works. It's proving to be
difficult as Microsoft never really provided a way of getting taskbar
positioning to work... I've had to reverse engineer how taskbar placement
works, and it's been pretty tricky to do.

But once I have something weighing properly on my test systems then I'll
get you to test again for me.

Thanks
Terry

**Comment by RBZL** [comment]
> What do you think? Does having Profile Tools send a mixed message?

I think it's neat that it works, but the number of clicks it takes me to get there takes longer than unlocking and moving my taskbars around. I'd be more likely to move them myself than ever use the Profile Tools way, and I don't expect needing to move them often anyway. The only time I touch them is when they get messed up somehow.

Let me know when you've got something new to test and I'll check it out.

**Comment by terrymacdonald** [comment]
@RBZL, I've been deep in the internals of Windows trying to set windows taskbar locations using windows internal messages for the last 4 weeks. It's not worked as I want to, so I'm pulling the Windows Taskbar setting code from the upcoming release of DisplayMagician v2.2.0.

To be clear, any new Display Profiles created in the upcoming DisplayMagician v2.2.0 will store the taskbar layout information in them, but the code that applies the windows taskbar layouts has been disabled. This is because it was not working reliably. 

As you may know, Windows 10 and Windows 11 don't have any API based ways to move windows taskbars about. There is a registry key way (which has a 5 second delay  so won't work), or there is a mouse based drag/drop way (which doesnt work in Windows 11). As such I've been trying to develop my own way of moving the taskbars directly!

I've been trying to reverse engineer the taskbar dragging/dropping internal windows messages to see if I can fool windows into thinking that the taskbars have been moved by the user. I have managed to get the taskbars to move to the locations they are supposed to be, **but the start menu doesn't move along with the taskbar**. Even worse, the start menu stops opening completely until the next PC reboot. This is bad.

It's REALLY annoying, but I will have to drop the taskbar functionality from DisplayMagciain v2.2.0 as it is not reliable enough, and I will just get lots of bug reports and complaints from users. I am still trying to figure out the missing bits to see if I can finally get things to work, but for now I have to work on getting this release out the door. DisplayMagciain v2.2.0 has a lot of other bug fixes in it, and it will stop a lot of the problems people are logging issues for now. 

So, we will see how things go, and hopefully I'll find the magic setting that makes taskbar setting work. 🤞 

Thanks
Terry

Just wanted to keep you updated.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@RBZL Can you please help?

I've been working the last month to add in the GetDisplayConfig and SetDisplayConfig support into the NVIDIALibrary, which grabs a lot more NVIDIA settings than I used to get. But I need people with different hardware to test things for me. I need to test that the new NVIDIALibrary updates I made will get your hardware working properly.

I need you to test NVIDIAInfo, a little app that I wrote specifically to test the NVIDIALibrary that I wrote. Because I want to only test the video manipulation works with multiple video cards, we will only use that for testing this. Once we can confirm that it all works then I'll roll it into DisplayMagician.

Can you please download and extract this NVIDIAInfo zip file to a folder on your computer:
[NVIDIAInfo-v1.7.0.zip](https://github.com/terrymacdonald/NVIDIAInfo/releases/tag/v1.7.0)
 
Open a PowerShell window, and change to the same folder that you extracted NVIDIAInfo into.

**IMPORTANT - These instructions are likely to fail, so I need you to save/rename the NVIDIAInfo.log file as shown below. Each run of NVIDIAInfo overwrites the previous NVIDIAInfo.log, so I need you to rename it each time so it doesn't get overwritten. I need to see the log file entries to know what is broken.**

Please complete the following instructions in this specific order:

Create a surround display layout with some of your screens then:
- Save the first config as `NVIDIAInfo save surround.cfg`
- Rename the NVIDIAInfo.log file to `Rename-Item -Path "NVIDIAInfo.log" -NewName "NVIDIAInfo-surround-save.log"`
- Check if the first windows display config is equal with `NVIDIAInfo equal surround.cfg` (we want it to be equal)

Next create a second windows display layout, this time with individual triple screen:
- Save the second windows display as `NVIDIAInfo save triple.cfg`
- Rename the NVIDIAInfo.log file to `Rename-Item -Path "NVIDIAInfo.log" -NewName "NVIDIAInfo-triple-save.log"`
- Check if the second windows display config is equal with `NVIDIAInfo equal triple.cfg` (we want it to be equal)
- Check if the first windows display config is NOT equal with `NVIDIAInfo equal surround.cfg` (we want it to NOT be equal)

Next create a third windows display layout, this time with a cloned display in it:
- Save the second windows display as `NVIDIAInfo save cloned.cfg`
- Rename the NVIDIAInfo.log file to `Rename-Item -Path "NVIDIAInfo.log" -NewName "NVIDIAInfo-cloned-save.log"`
- Check if the third windows display config is equal with `NVIDIAInfo equal cloned.cfg` (we want it to be equal)
- Check if the second windows display config is equal with `NVIDIAInfo equal triple.cfg` (we want it to NOT be equal)
- Check if the first windows display config is NOT equal with `NVIDIAInfo equal surround.cfg` (we want it to NOT be equal)

Next swap back to the surround display layout:
- Load the first windows display as `NVIDIAInfo load surround.cfg`
- Rename the NVIDIAInfo.log file to `Rename-Item -Path "NVIDIAInfo.log" -NewName "NVIDIAInfo-surround-reload.log"`
- Check if the first windows display config is equal with `NVIDIAInfo equal surround.cfg` (we want it to be equal)
- Save a copy of the current display config for me to compare: `NVIDIAInfo save surround-reload.cfg`
- Make sure the second windows display config is NOT equal with `NVIDIAInfo equal triple.cfg` (we DO NOT want it to be equal)

Next swap back to the triple screen display layout again:
- Load the second windows display as `NVIDIAInfo load triple.cfg`
- Rename the NVIDIAInfo.log file to `Rename-Item -Path "NVIDIAInfo.log" -NewName "NVIDIAInfo-triple-reload.log"`
- Check if the second windows display config is equal with `NVIDIAInfo equal triple.cfg` (we want it to be equal)
- Save a copy of the current display config for me to compare: `NVIDIAInfo save triple-reload.cfg`
- Make sure the first windows display config is NOT equal with `NVIDIAInfo equal surround.cfg` (we DO NOT want it to be equal)

Next swap back to the cloned screen display layout again:
- Load the second windows display as `NVIDIAInfo load cloned.cfg`
- Rename the NVIDIAInfo.log file to `Rename-Item -Path "NVIDIAInfo.log" -NewName "NVIDIAInfo-cloned-reload.log"`
- Check if the second windows display config is equal with `NVIDIAInfo equal cloned.cfg` (we want it to be equal)
- Save a copy of the current display config for me to compare: `NVIDIAInfo save cloned-reload.cfg`
- Check if the second windows display config is equal with `NVIDIAInfo equal triple.cfg` (we want it to be equal)
- Make sure the first windows display config is NOT equal with `NVIDIAInfo equal surround.cfg` (we DO NOT want it to be equal)

Once that series of tests are done, please include the following files in a ZIP file, and upload the ZIP to this github issue through your web browser (click and drag the file from your File explorer into Github and it will upload):
- surround.cfg
- surround-reload.cfg
- NVIDIAInfo-surround-save.log
- NVIDIAInfo-surround-reload.log
- triple.cfg
- triple-reload.cfg
- NVIDIAInfo-triple-save.log
- NVIDIAInfo-triple-reload.log
- cloned.cfg
- cloned-reload.cfg
- NVIDIAInfo-cloned-save.log
- NVIDIAInfo-cloned-reload.log

This will give me all the information I need to know whether it worked! I need to have some confirmation from people with complex hardware so that I can be sure that it will work when the new version of NVIDIALibrary is integrated back into DisplayMagician. 

I really hope it works as this has been a TON of work :)

Thanks
Terry

**Comment by RBZL** [comment]
I'm preparing for a trip in a few days so I may not have time to test it out prior to leaving, but I'll be back around April 4 and can try to check it out that week.

**Comment by RBZL** [comment]
Got around to this today, here you go:

[NVIDIAInfo-tests.zip](https://github.com/terrymacdonald/DisplayMagician/files/8457855/NVIDIAInfo-tests.zip)

Pretty slick, and the taskbars behaved too. Not sure if that was coincidence or not. I didn't move the taskbars around all crazy in my tests, but the center taskbar returned to the center screen when I activated triple.cfg.

Notes:

I encountered the following exception a few times:

```
NVIDIAInfo v1.7.0
=================
By Terry MacDonald 2022

Unhandled Exception: System.TypeInitializationException: The type initializer for 'DisplayMagicianShared.NVIDIA.NVIDIALibrary' threw an exception. ---> System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at DisplayMagicianShared.NVIDIA.NVImport.NvAPI_DISP_GetDisplayConfig(UInt32& PathInfoCount, NV_DISPLAYCONFIG_PATH_INFO_V2[]& PathInfos, Boolean thirdPass) in H:\vscode-projects\NVIDIAInfo\NVIDIAInfo\NVAPI.cs:line 4008
   at DisplayMagicianShared.NVIDIA.NVIDIALibrary.GetNVIDIADisplayConfig(Boolean allDisplays) in H:\vscode-projects\NVIDIAInfo\NVIDIAInfo\NVIDIALibrary.cs:line 777
   at DisplayMagicianShared.NVIDIA.NVIDIALibrary.GetActiveConfig() in H:\vscode-projects\NVIDIAInfo\NVIDIAInfo\NVIDIALibrary.cs:line 338
   at DisplayMagicianShared.NVIDIA.NVIDIALibrary..ctor() in H:\vscode-projects\NVIDIAInfo\NVIDIAInfo\NVIDIALibrary.cs:line 212
   at DisplayMagicianShared.NVIDIA.NVIDIALibrary..cctor() in H:\vscode-projects\NVIDIAInfo\NVIDIAInfo\NVIDIALibrary.cs:line 170
   --- End of inner exception stack trace ---
   at DisplayMagicianShared.NVIDIA.NVIDIALibrary.GetLibrary()
```

H: is a mapped network drive to a RPi on my network so I'm not sure why NVIDIAInfo was trying to do anything with that path. It happened 5 times while attempting to perform various steps in your instructions. Re-running the command afterwards resulted in success each time. It happened after these commands:

```
.\NVIDIAInfo save cloned.cfg
.\NVIDIAInfo equal surround.cfg
.\NVIDIAInfo equal cloned.cfg x2
.\NVIDIAInfo load triple.cfg
```

Since it happened while saving the cloned.cfg, I continued with that section of the instructions. Then, I repeated the cloned section instructions which saved the config successfully, and renamed the log file -2 so there will be two logs for cloned-save - the first one might be weird, but the second one is after a successful run.

Let me know if you need anything else.

**Comment by terrymacdonald** [comment]
Hi @RBZL,

As I mentioed before I've had to remove the custom Windows taskbar setting functionality from upcoming the DisplayMagician v2.3.0 release, but I'd love it if you could test that release for me. DisplayMagician v2.3.0 fixes a lot of issues (including the Unhandled exception errors you posted above).

I'm hoping this version is reliable enough to be the next long term release. We'll see how that goes based in the testing everyone is doing for me :).

NVIDIALibrary can now handle multiple physical adapters in a logical group, dvi-d and earlier connections, USB connected displays, adaptive sync, image scaling, cloned displays and more. 

[DisplayMagicianSetup-v2.3.0.35.zip](https://github.com/terrymacdonald/DisplayMagician/files/8457813/DisplayMagicianSetup-v2.3.0.35.zip)

Can you please download and install DisplayMagician v2.3.0.35 from the link above, and then do the following two steps:

- Delete your old Display Profiles 
- Create new Display Profiles 
- Update your Game Shortcuts to use the new Display Profiles

If you have any issues, please create a Support ZIP File (main screen > Settings > Create Support ZIP File) and upload it here through your web browser. I'll then be able to use that for any bug fixing.
 
Thanks
Terry

**Comment by terrymacdonald** [comment]
Sorry! We just found an issue with some botched Display Profile upgrade logic :(.  Now fixed.

Please do the following:

- Delete `C:\Users\<Username>\AppData\Local\Displaymagician\Profiles\DisplayProfiles_2.3.json` if it exists.
- Install this new DisplayMagician version:
[DisplayMagicianSetup-v2.3.0.36.zip](https://github.com/terrymacdonald/DisplayMagician/files/8467396/DisplayMagicianSetup-v2.3.0.36.zip)
- Run DisplayMagician
- Create your Display Profiles again the way you want them
- Update your Game Shortcuts to use the new Display Profiles
- Have fun with your games!

Hopefully it should work this time.  Please log a new issue at https://github.com/terrymacdonald/DisplayMagician/issues/new/choose if you experience any new bugs that you've not seen before, and I'll get right on fixing them.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@HS34 I think this version will work for you. Can you please install this version instead of the one I suggested in previous comments and test that the problem is fixed, and then report back here.
[DisplayMagicianSetup-v2.3.1.7.zip](https://github.com/terrymacdonald/DisplayMagician/files/8494408/DisplayMagicianSetup-v2.3.1.7.zip)

If the issue is fixed then please also close this issue so I can keep track of when I can release this new version to the general population in the next few days.

Thanks
Terry

**Comment by HS34** [comment]
Hi Terry,
I can't view current display with this version (same problem with 2.3.0.50)
![image](https://user-images.githubusercontent.com/24624829/163587997-9082958e-03ab-4bc6-a19c-2420d43db01b.png)

When saving the profile, I've got the following exception :
[Exception.txt](https://github.com/terrymacdonald/DisplayMagician/files/8496658/Exception.txt)
 
Associated log :
[DisplayMagician.log](https://github.com/terrymacdonald/DisplayMagician/files/8496640/DisplayMagician.log)

Regards,
Hervé

**Comment by terrymacdonald** [comment]
Hi,

Can you please create a  for me? Click on the Settings button on the main
DisplayMagician window, then click "Create Support ZIP File" button. Please
upload that file here as it gives me all the files I need.

Thanks
Terry

**Comment by HS34** [comment]
Sure :
[DisplayMagician-Support-20220415-2201.zip](https://github.com/terrymacdonald/DisplayMagician/files/8498041/DisplayMagician-Support-20220415-2201.zip)

**Comment by terrymacdonald** [comment]
OK @HS34 It looks like it's something to do with having two video cards that is causing the issue. I'll have to do some detective work to try and figure out what is broken in order to figure out a fix. I only have one video card so it's not something I can test. DisplayMagician is breaking just before it writes the display config, which means I can't see your display layout to understand what is causing DisplayMagicain to break :(. 

But - I have an idea. In order to get an understanding if your configuration, can you please download NVIDIAInfo v1.7.2 from here: [https://github.com/terrymacdonald/NVIDIAInfo/releases/tag/1.7.2](https://github.com/terrymacdonald/NVIDIAInfo/releases/tag/1.7.2)

Open a PowerShell window, and change to the same folder that you extracted NVIDIAInfo into.

Then please complete the following instructions in this specific order:

Using the same display layout that caused the error in DisplayMagician, please save the display layout:
- Save the first config as `NVIDIAInfo save broken.cfg`
- Rename the NVIDIAInfo.log file to `Rename-Item -Path "NVIDIAInfo.log" -NewName "NVIDIAInfo-broken-save.log"`
- Check if the saved display config is equal with `NVIDIAInfo equal broken.cfg` (we want it to be equal)

Once that test is done, please include the following files in a ZIP file, and upload the ZIP to this github issue through your web browser (click and drag the file from your File explorer into Github and it will upload):
- broken.cfg
- NVIDIAInfo-broken-save.log

**What we're trying to do**
We're trying to record the troublesome display layout into a cfg file so that I can see exactly what your display layout is. I will then be able to use that cfg file to figure out what bits are broken in the DisplayMagician logic that is causing the issue for you. Once I understand what is wrong I can write new code that will avoid the issue, and then everything will work again and there'll be unicorns and fireworks and happiness.

Thanks
Terry

**Comment by terrymacdonald** [comment]
Also, I've tried to preempt some potential problem points, and I've made some changes that _may_ help fix the problem.

Can you please also install DisplayMagician v2.3.1.12 which I've attached here:
[DisplayMagicianSetup-v2.3.1.12.zip](https://github.com/terrymacdonald/DisplayMagician/files/8499167/DisplayMagicianSetup-v2.3.1.12.zip)
 
It might work now.... Please tell me if it works now or if there is an error, and please also do the NVIDIAInfo tasks as per my last comment.

Thanks!
Terry

**Comment by terrymacdonald** [comment]
@hs34 @RBZL I have worked through many issues over the last week, so I recommend that you do some testing with this version:
[DisplayMagicianSetup-v2.3.1.50.zip](https://github.com/terrymacdonald/DisplayMagician/files/8512144/DisplayMagicianSetup-v2.3.1.50.zip)

I **strongly** recommend that  you delete your old display profiles and recreate them again just to be 100% sure you're running the latest versions of the file, and then updated your game shortcuts to use those new profiles.

This version fixes all the bugs that I know about, and based on my testing is just about ready to release to the general public. I'd really appreciate it if you could do some testing (but only if you have some spare time) so that I will know if it really does fix all the issues.

Thanks
Terry

**Comment by HS34** [comment]
Hi,
Thanks Terry, this version works great ! 
All switches ok (4 screens, surround, cloned). That's the most important for me. Thank you.

Remaining problems : 
- DisplayMagician just exited after the first (successfull) switch from surround to "4 screens". See DisplayMagician1.log. No problem with following switches.
- Windows task bar position is not restored (it should be on my right screen, but it was restored on the middle one).
- DisplayMagician doesn't see that I moved the Windows task bar, even if I clic on "View Current Display".
[DisplayMagician-Support-20220420-0843.zip](https://github.com/terrymacdonald/DisplayMagician/files/8518952/DisplayMagician-Support-20220420-0843.zip)

Regards,
Hervé

**Comment by terrymacdonald** [comment]
@HS34 I've had to remove the taskbar placement support from DisplayMagician. It wouldn't reliably work. I managed to move the taskbars using some tricky windows manipulation logic, but it meant that the start menu wouldn't open at all. Which is not what people want :).

I therefore decided to park the taskbar display logic for another 6 months or so, as I've heard rumors that Microsoft will be adding some more logic to allow taskbars to be moved in Windows 11. I'm parking that code until then.

DisplayMagician exiting just after a switch is not good news at all. There were unfortunately no errors in the logs that could help me. They just stopped near the point it crashed. This means there isn't really anything I can do at the moment. I guess we'll have to keep an eye on it and if there is still a problem we'll see if we can find the issue (though not sure how).

The last two errors you've described won't be fixed until I restart work on the taskbar code in 6 months or so.

At this point, there's not much I can do about taskbar position, other than to hope that Windows and the NVIDIA driver start tracking taskbar position when there are future updates.

**Comment by HS34** [comment]
Thanks Terry,
Taskbar position is not a big problem. It's fast to move it after the switch. And you're right... I need the start menu from time to time... ;)
I won't move to windows 11 on this PC (too old). I will keep it as is to play old games in 3D surround as long as I can.
DisplayMagician is a great tool !

**Comment by terrymacdonald** [comment]
Yes it's like to hear about what you found out. It might help me
troubleshoot the code I wrote.

Thanks
Terry

**Comment by sk666** [comment]
ok this will be a long post, but the solution take 5-10min to do. @HS34 and @RBZL just jump at the end of this post to have the solution without the rest of the blablabla. I'm sorry for my english, it's not my first language and I'm writting this without google translate, I hate translators.

* My computer: 
w10 22h2
Nvidia RTX3060
SSD drive (not NVMe)
Three of the exact same Dell monitor, all connected with DP cable

* Display default settings: Same as RBZL user: 
Tripple display, middle screen is the primary display:

![01 DM profiles](https://github.com/terrymacdonald/DisplayMagician/assets/29050341/a22a39bd-3cf9-492e-b836-42386241bc1a)
![02 nvidia display settings before surround](https://github.com/terrymacdonald/DisplayMagician/assets/29050341/9b3986d1-7168-4ca3-bb5d-a3cbe66d96d7)
![03 display settings before surround](https://github.com/terrymacdonald/DisplayMagician/assets/29050341/29cd45f7-26d3-4b7e-94a3-8496b324f76e)

* Behavior: 
the same as RBZL user: When turning off surround, primary display taksbar (the one with system tray, date, clock and my customizez quick launch section) move to my right-most monito. Surprinsingly, windows detect that my middle monitor is still my primary monitor. 

* What I thought: 
if DM and/or windows and/or nvidia can decide to move taskbar position and wrongfully detect my primary monitor to be my right-most monitor instead of my middle one monitor after turning off surround, may be I can work with primary display monitor detection instead moving the taskbar position.

So, what I did on my old rig, who have NOT a fresh windows installation:

1- download this tool: NirCmd v2.86 x64 (http://www.nirsoft.net/utils/nircmd.html)
2- extract it
3- create a text file in this previous extracted zip folder to make a script
4- I wrote this in it:

`nircmd.exe setprimarydisplay 1
Timeout /T 4
nircmd.exe setprimarydisplay 3
Timeout /T 4
nircmd.exe setprimarydisplay 2`

5- rename the extension of the text file to .bat
6- open DM and go to Game Shortcut section
7- edit the shortcut you want to test
8- go to tab "5. Choose what happens afterwards"
9- check the box "Run this program"
10- click on the "Choose" button
Note: at this point, DM will only see .exe and .com file
11- click on the right section 'exe files (*.exe,*.com) and choose "All files"
12- navigate to the .bat file of the #5 previous step
13- click on "Open" button
14- save and close DM shorcut edition
15- test it.

* Now, What I did and why:
I did extensive test here to find the best scenario. If I select only 2 display monitor one the code instead of 3 monitor, it does not work whatever which I take. Also, If I choose a 1 second timer, it did nothing. If I choose a 2 second timer, the taskbar will return to the middle monitor like I want it (Yayyy!) but in a vertical position at the right of the middle (damn it!). If I choose a 3 second timer, sometimes it's working well, sometimes it will do the same as the 2 second timer. 4 second timer always worked on my test, but may be it will take more time on other user computer. I decided to beginto start with primary display #1 because the "primary taskbar" is always on this monitor when I turn off surround. I wrote the code without echo-off for debugging purpose to be able to see what it did and when he did it, so I supposed you can add echo-off if you want.

It will be interesting to test it with a NVMe SSD (and/or with w11) to see if the timeout could be less that 4 second after each primary display switch. I am unable to do theses tests right now. 

Please note that i'm just a pro it-technician on my day to day, I'm not a programmer and I make code only for the fun without programming knowledge. You far better than me Terry and may be you will find a better way to do it or doing it without this tool. 

I don't fully understand why taskbar position broke with your Nvidialibrary. I don't fully understand what is the culprit between Nvidia or Microsoft, but if I I'm not mistaken, I saw AMD user on the net who seem to have the same task bar problem after using eyefinity. But I can be wrong for AMD users. I hope that will help you to "fix" this, DM v2 is far better than the v1, I cross my fingers that DM v2 or a futur DM release will be handle this taskbar problem with w11 like the DM v1 did with w10.

Final note: I know, I put a lot of time in this just to save a couple of seconds tweek after turning off surround. I will be honest, I'm lazy but I hate doing repetitive task everyday where and when I can make them automatically.



For @HS34 and @RBZL, here is my workaround:

1- download this tool: NirCmd v2.86 x64 (http://www.nirsoft.net/utils/nircmd.html)
2- extract it
3- create a text file in this previous extracted zip folder to make a script
4- I wrote this in it:

`nircmd.exe setprimarydisplay 1
Timeout /T 4
nircmd.exe setprimarydisplay 3
Timeout /T 4
nircmd.exe setprimarydisplay 2`

5- rename the extension of the text file to .bat
6- open DM and go to Game Shortcut section
7- edit the shortcut you want to test
8- go to tab "5. Choose what happens afterwards"
9- check the box "Run this program"
10- click on the "Choose" button
Note: at this point, DM will only see .exe and .com file
11- click on the right section 'exe files (*.exe,*.com) and choose "All files"
12- navigate to the .bat file of the #5 previous step
13- click on "Open" button
14- save and close DM shorcut edition
15- test it.

I hope this workaround will fix this problem just it did for me.

**Comment by sk666** [comment]
Also, that is intesresting that me and RBZL user wanted to have the primary taskbar with sys tray on the middle monitor and, insted of this, we had it on the right-most monitor while  HS34 user wanted to had it on the right side monitor and he had it on the middle monitor... we want the behavior of HS34 user and he want ours.

I wounder if someone would want it to the left side monitor (in a tripple display setup),will the taskbar go in the middle monitor?

**Comment by terrymacdonald** [comment]
There is a lot to unpack here @sk666. It will take me quite a lot of time
to go through the information you have provided. Thank you for putting this
information together. I will use it for further testing in a few weeks when
I get v2.6.0 out the door.

I would like to clarify that DM v1 didn't have any taskbar specific code in
it at all. The taskbar position was set most likely by Windows (in triple
screen or single screen mode), or by NVIDIA (in surround mode). As far as I
can tell the taskbar position was set by Windows when it recognised the
screen layout.

I spent about a month experimenting with trying to record and reset the
taskbar position using programming. I tried interacting with the windows
registry, I tried sending various windows messages directly to the windows
subsystems, I even tried manually manipulating the taskbar windows
themselves. Nothing worked reliably. After 4 weeks of trying I parked the
idea, but I did leave in the taskbar layout recording information in
DisplayMagician. You can see that if you open a DM dispay_profiles.json
file - there is a section in each profile outlining the location of the
taskbars. The reason I left that in was if I managed to find a way to
reliably move the taskbars in the future the data would be there to tell me
where it needed to be moved to.

I have never been able to find example code that shows me how to manipulate
the taskbar position. I may be able to build in some taskbar manipulation
code if I can find some existing code to work. I don't know enough about
how the Windows taskbar ui manipulation works in order to do it myself. So
I'm happy you found nircmd. I'll try taking a look at the codebase to see
if there is anything that could help.

Thanks

*Terry*

**Comment by terrymacdonald** [comment]
ok @sk666 - looking through the nircmd commands you've listed, it's using the setprimary display to force windows to draw the taskbar in the right place. That is a very interesting idea. I do actually record which screen is the primary one, and that info is stored as part of the windows configuration information that we get from both the Windows CCD and the older Windows GDI interface. 

Interestingly nircmd seems to use the older GDI interface for setting the primary display, as I recognise the setprimarydisplay structure. 

Your question did set me on the path to finding the following bit of code here: https://stackoverflow.com/questions/195267/use-windows-api-from-c-sharp-to-set-primary-monitor#195272

In this, they use the GDI method ChangeDisplaySettingsEx() to set the primary monitor by making it's coordinates 0,0, and then setting the rest of the monitors to the same offset, effectively moving the (0,0) point to be the top of the monitor you want to be the primary monitor. Then calling the ChangeDisplaySettingsEx()  one final time to set the final location.

My code already should be doing this, as I use ChangeDisplaySettingsEx() that same way, and if the primary monitor is set correctly then that GDI change should allow it to work to reset the (0,0) point. It is possible I have a bug in that part of the codebase though, so it may not be settings the screen correctly.

**It's interesting that it seemed to work reliably when it was done a second time.** I might be able to add in some code to check if the primary screen was actually changed, and then if not to repeat the GDI screen move a second time to see if that 2nd attempt will actually make the change. That may enable everything to work correctly.

In any case you've given me some nice new tests to do, so I'll see what I can do over the next few days to test if that idea works or not. If I get something working I'll paste it in here for you to test.

Thanks
Terry

**Comment by sk666** [comment]
Oooh that's sound great :) 

I don't know if this will give you some hint but I discovered something by accident yesterday when I tested your .492 build. Yesterday jjst for a testing purpose, I used the DM context menu (right click menu) on the desktop to switch to surround and to return to my default display.

I'm not totally sure why but when I switch to surround by the context menu, the taskbar seem to stay in the good primary monitor. In the day to day when I used DM to turn on and turn off surround, I used a game desktop icon/taskbar shortcut. 

The context menu only switch to surround without any other actions so may be you can find the bug if you look on the "context menu code". Or may be not but it's interesting.

**Comment by sk666** [comment]
Did you change something about what I said in my last comment on the build .492? The context menu was replaced the primary taskbar in the right place but .446 not doing it today. So if you did not change anything, I think I was lucky yesterday, as well as 1 day in december 2023 when I tested to remove hidden display in the device manager trying to fix a weird glitch on my dell monitor (I  did not have this glitch with my previous old monitors). Whatever, when I removed the 2 hidden display in device manager,  I rebooted my computer and I played a game on steam with display magician and it was the only time DM was able to replace the primary taskbar in the right place. So I was may be lucky that day and yesterday. And at the tine, it was with the last official build DM 2.5.x.x .

That is sooo weiiird.

**Comment by terrymacdonald** [comment]
Hi @sk666,

No I haven't changed anything on your system. I don't have access to any
systems of any users - the most I can do is tell the update server that
there is a new update available, and then DM will tell the user they should
update. The user still has to click the update button to make that happen.

I've got no idea why it worked that other time and not this one. I'm
wondering if it's some kind of 'race condition' where some parts of windows
are taking time to update themselves after the change happens, and that is
causing problems as DM attempts to change the monitor positions. For this
reason I've added a 0.5s delay into the codebase between when the
WinLibrary uses the CCD API and GDI API to set the monitor layouts. I've
also tweaked the ChangeDisplaySettingsEx DEVMODE structure that's used to
hopefully be a bit more reliable in how it's being used. I'm hoping that
results in Windows knowing the primary display better, and that then
helping Windows put all the taskbars back in the same place. We shall see
when I release the next test version of DM for you to trial.

Thanks
Terry

**Comment by sk666** [comment]
Sorry I mean if you changed someting in the code, not in my desktop haha.

Thanks for your work on the additional delay, not sure if this gona do the trick or if we absolutely need to switch the primary display 2 times for a tripple display setup. We will be fixed soon.

**Comment by terrymacdonald** [comment]
@RBZL The latest test version should be working properly, and hopefully fixes your issue. Could you please test it for me so I know that it fixes your issue? If it does then I'll be able to close this issue as the fix will be part of the next full release.

[DisplayMagicianSetup-V2.6.0.515-No-NvApi_SetDisplayConfig.zip](https://github.com/terrymacdonald/DisplayMagician/files/14200905/DisplayMagicianSetup-V2.6.0.515-No-NvApi_SetDisplayConfig.zip)

As I've not really had any feedback from you recently I'll give it a week and then close this issue if I have no further responses. We can always reopen this issue in the future if need be.

Thanks!

**Comment by RBZL** [comment]
I'll try to check it out within the next few weeks. I'll let you know
either way!

**Comment by terrymacdonald** [comment]
Oops I broke the setup with my previous version.  This version has a corrected setup that should install everything properly.

[DisplayMagicianSetup-v2.6.0.522-corrected-setup.zip](https://github.com/terrymacdonald/DisplayMagician/files/14249946/DisplayMagicianSetup-v2.6.0.522-corrected-setup.zip)

Thanks
Terry

**Comment by RBZL** [comment]
Finally got around to testing this. Taskbars appear to be staying where they're supposed to between profile changes.

I found a separate issue where DM isn't retaining the refresh rate on a 4th screen (TV) connected via HDMI. It's set to 1920x1080 120Hz when the profile is saved, but recalls as 60Hz when the profile is loaded. I can see if there are other issues for this or create a new issue if needed.

**Comment by terrymacdonald** [comment]
Hi RBZL,

Are those refresh rates set in the NVIDIA driver, or within windows? I've
currently disabled the NVIDIA SetDisplayConfig function within DM and that
has probably stopped the application of refresh rate settings if they're
defined within NVIDIA.

I'm rapidly coming to the conclusion that disabling SetDisplayConfig won't
work. I think I'll instead need to repair that function within my
NVIDIALibrary. I'm going to look at how Soroush Falahati handled that part
within NvAPIWrapper and probably copy that part of his work. Soroush is an
amazing developer and he solved this issue a long time ago. He has way more
knowledge than I do in this area, so I should definitely crib from his work.

I think I might start work on fixing the NVIDIALibrary SetDisplayConfig
shortly. I'll also think there are a couple more mosaic topology functions
I need to add as well, so I might add will add those too!

This will all take a little while to sort out, but I really think it's
needed to ensure I cover the most commonly used NVIDIA settings.

Thanks
Terry

**Comment by RBZL** [comment]
@terrymacdonald
I did some testing, and it doesn't matter where I set the refresh rate, the NVidia Control Panel or Windows' advanced display settings. The 120Hz refresh rate isn't being recalled by DM.

**Comment by terrymacdonald** [comment]
Hmmmm. That is odd.

I'm currently rewriting the NVIDIALibrary to use Soroush Falahati's code as
he writes way better choice than I do, and I'm planning on adding back
SetDisplayConfig into DisplayMagician which should take care of that issue
from the NVIDIA side.

It's odd that the refresh rate within the windows side isn't working
though. That should work fine! I've not made any changes there.... So it
looks like I have some more work to do there and will.

Thanks
Terry

**Comment by RBZL** [comment]
Sounds good. I'll close this issue and make a new one for that for tracking. Thanks for following up on this one!
